compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java

```diff
@@ -0,0 +1,1906 @@
+
+        clone.indexExpr = clone(source.indexExpr);
+        cloneBLangAccessExpression(source, clone);
+    }
+
+    private void cloneBLangStructureTypeNode(BLangStructureTypeNode source, BLangStructureTypeNode clone) {
+
+        clone.fields = cloneList(source.fields);
+        clone.initFunction = clone(source.initFunction);
+        clone.isAnonymous = source.isAnonymous;
+        clone.isFieldAnalyseRequired = source.isFieldAnalyseRequired;
+        clone.typeRefs = cloneList(source.typeRefs);
+    }
+
+    private void cloneBLangMatchBindingPatternClause(BLangMatch.BLangMatchBindingPatternClause source,
+                                                     BLangMatch.BLangMatchBindingPatternClause clone) {
+
+        clone.body = clone(source.body);
+        clone.matchExpr = clone(source.matchExpr);
+        clone.isLastPattern = source.isLastPattern;
+    }
+
+    private void cloneBLangType(BLangType source, BLangType clone) {
+
+        clone.nullable = source.nullable;
+        clone.grouped = source.grouped;
+    }
+
+    private <T extends Enum<T>> EnumSet<T> cloneSet(Set<T> source, Class<T> elementType) {
+
+        if (source == null || source.isEmpty()) {
+            return EnumSet.noneOf(elementType);
+        } else {
+            return EnumSet.copyOf(source);
+        }
+    }
+
+    /* Visitor Methods */
+
+    @Override
+    public void visit(BLangPackage pkgNode) {
+        // Ignore.
+    }
+
+    @Override
+    public void visit(BLangTestablePackage testablePkgNode) {
+        // Ignore.
+    }
+
+    @Override
+    public void visit(BLangCompilationUnit source) {
+
+        BLangCompilationUnit clone = new BLangCompilationUnit();
+        source.cloneRef = clone;
+        clone.name = source.name;
+        for (TopLevelNode node : source.topLevelNodes) {
+            clone.topLevelNodes.add(clone(node));
+        }
+    }
+
+    @Override
+    public void visit(BLangImportPackage source) {
+
+        BLangImportPackage clone = new BLangImportPackage();
+        source.cloneRef = clone;
+        clone.pkgNameComps = source.pkgNameComps;
+        clone.version = source.version;
+        clone.alias = source.alias;
+        clone.orgName = source.orgName;
+        clone.compUnit = source.compUnit;
+    }
+
+    @Override
+    public void visit(BLangXMLNS source) {
+
+        BLangXMLNS clone = new BLangXMLNS();
+        source.cloneRef = clone;
+        clone.namespaceURI = clone(source.namespaceURI);
+        clone.prefix = source.prefix;
+    }
+
+    @Override
+    public void visit(BLangFunction source) {
+
+        BLangFunction clone = new BLangFunction();
+        source.cloneRef = clone;
+
+        clone.attachedFunction = source.attachedFunction;
+        clone.objInitFunction = source.objInitFunction;
+        clone.interfaceFunction = source.interfaceFunction;
+        clone.anonForkName = source.anonForkName;
```

Comment:
Can't we move this into `cloneBLangInvokableNode` as well?

Reply:
Design is to keep attribute clone with relevant visit method. If attribute extends from the superclass,  those are moved to the "clone<SupperClasName>" method.  In this case, these attributes are coming from BLangFunction node, hence kept with BLangFunction node visit method.